From 86e26019938679b5aa60d258ab52b7c3009d0008 Mon Sep 17 00:00:00 2001
From: Balamurugan Ramar <quic_bramar@quicinc.com>
Date: Sat, 30 Mar 2024 19:02:27 +0530
Subject: [PATCH 2/2] UPSTREAM: mac80211: Fix for use-after-free in key memory

Currently while stopping a link AP, all keys from the interface were
removed. However with MLO there is a requirement to free only the link
keys.

Add changes to remove keys which are associated with the link AP which
is going to be stopped.

(Cherry-picked from
https://patchwork.kernel.org/project/linux-wireless/patch/20240227042251.1511122-2-quic_adisi@quicinc.com/)

Signed-off-by: Rameshkumar Sundaram <quic_ramess@quicinc.com>
Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
Signed-off-by: Balamurugan Ramar <quic_bramar@quicinc.com>
---
 net/mac80211/cfg.c                    |  7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index b526026..db755fb 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1801,6 +1801,7 @@ static int ieee80211_stop_ap(struct wiphy *wiphy, struct net_device *dev,
 	struct ieee80211_link_data *link =
 		sdata_dereference(sdata->link[link_id], sdata);
 	struct ieee80211_bss_conf *link_conf;
+	LIST_HEAD(keys);
 
 	if (WARN_ON(!link))
 		return -ENOLINK;
@@ -1889,6 +1890,12 @@ static int ieee80211_stop_ap(struct wiphy *wiphy, struct net_device *dev,
 
 	__sta_info_flush(sdata, true, link_id);
 
+	ieee80211_remove_link_keys(link, &keys);
+	if (!list_empty(&keys)) {
+		synchronize_net();
+		ieee80211_free_key_list(local, &keys);
+	}
+
 	ieee80211_stop_mbssid(sdata, link_id);
 	link_conf->enable_beacon = false;
 	link_conf->mbssid_tx_vif = NULL;
-- 
2.34.1

